### PR TITLE
[stdlib] Avoid malloc_size on OpenBSD.

### PR DESCRIPTION
**Pull request by 3405691582**
malloc introspection is a platform feature that is unavailable on
OpenBSD. There is no workaround for the feature, so simply avoid calling
it in this case. It is assumed on OpenBSD that the tail allocator will
allocate the amount of memory requested of it, and nothing more.

### CODE REVIEW

**Comment by 3405691582** [comment]
cc @milseman who was working in this area recently

**Review by CodaFi — CHANGES_REQUESTED** [review]
Could you please add a comment here explaining what you wrote in the PR's summary?

**Comment by milseman** [comment]
Array also calls `malloc_size`, and has for years. String just recently added this. I'm curious, how did this manifest for you? Is there an alternative we can do in the shim instead?

**Comment by milseman** [comment]
CC @Catfish-Man @jckarter

**Comment by 3405691582** [comment]
> Could you please add a comment here explaining what you wrote in the PR's summary?

Done. I tweaked the language in the code comment to be a little more generic; who knows if there might be a porting effort by someone to another platform without malloc introspection.

> Array also calls malloc_size, and has for years. String just recently added this. I'm curious, how did this manifest for you? Is there an alternative we can do in the shim instead?

There are three uses of malloc_size in Swift that are problematic:

* ContiguousArrayBuffer.swift -- this is easy to fix, but this looks like it might be partially addressed in pr #30270, so it might just be easiest to wait for that.
* ManagedBuffer.swift -- this one is more complicated to fix, because the allocation and the introspection call contexts are separated far apart. I have an unpublished commit to kludge around this, but it's a bit of a hack.
* StringStorage.swift -- this pr.

I'm working on an OpenBSD port recently. It doesn't have malloc introspection, and introspection like this is something we can't kludge around, because knowing the allocation extents depends on introspecting the system allocator, so the only way around this is to track the size of the allocation in parallel with every allocation.

**Comment by jckarter** [comment]
It's going to be easy to forget about platforms like OpenBSD as we work on the standard library and refactor code, which might add new `malloc_size` checks in the future. It might be better for long term maintenance if we wrapped up all of the standard library's uses of `malloc_size` in a helper function that handles the case that it might be unavailable. We could add a `_stdlib_has_malloc_size` static inline function to the shims, and implement the helper like this:

```
internal func mallocSize(ptr: UnsafeRawPointer) -> UInt {
  if _stdlib_has_malloc_size() {
    return _stdlib_malloc_size(ptr)
  }
  return nil
}
```

That way, the conditional only has to be in one place for any OSes that don't support malloc introspection.

**Comment by 3405691582** [comment]
I'm happy to try and make that refactor here, that sounds simple and straightforward enough in itself, but I still don't know how to apply it for ManagedBufferPointer. That doesn't mean we can't at least _apply_ that scheme where it applies for the moment.

**Comment by jckarter** [comment]
@Catfish-Man or @milseman would be more familiar with the specific code at hand, but in general, you ought to be able to replace existing direct calls to `_stdlib_malloc_size(x)` with `mallocSize(x) ?? originalMallocArgument`, since we should be able to assume we got at least as much memory as we'd asked for originally.

**Comment by 3405691582** [comment]
Sure, but the problem with ManagedBufferPointer is that the malloc_size call is divorced from the allocation (see [_capacityInBytes](https://github.com/apple/swift/blob/master/stdlib/public/core/ManagedBuffer.swift#L434)), and ManagedBufferPointer [asserts](https://github.com/apple/swift/blob/master/stdlib/public/core/ManagedBuffer.swift#L396) on "illegal stored properties".

Happy to hear further advice on this, though.

**Comment by 3405691582** [comment]
Adopted @jckarter's suggestion to introduce a new shim. Still would like to hear if @Catfish-Man or @milseman have advice.

Points of note:
* as mentioned, ManagedBuffer.swift is not converted in this PR

* `mallocSize` shim returns `Optional<Int>` and not `UInt`, because the operators on pointers take `Int`s; the preexisting code also makes some assumptions based on `_swift_stdlib_malloc_size` returning an `Int`.

* the catchall implementation of `_swift_stdlib_malloc_size` returns 0 instead of asserting; it might make things a little too complex to try and drag in the necessary dependencies to get runtime assertions.

**Comment by compnerd** [comment]
Thanks for this, I think that the shim'ed approach is definitely nicer.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by 3405691582** [comment]
Looks like the OS X build failure is because swiftpm is being built with a platform Swift SDK for SDKROOT but with the Swift stdlib from HEAD + this PR, is that right? Is there something this PR is missing?

**Comment by CodaFi** [comment]
Lost the build failure

@swift-ci test macOS platform

**Review by benrimmington — COMMENTED** [review]
Suggestions:

* use `__swift_bool` if possible?

* and update the copyright year in all files.

  ```diff
  -// Copyright (c) 2014 - 20** Apple Inc. and the Swift project authors
  +// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
  ```

File: stdlib/public/SwiftShims/LibcShims.h
Comment (3405691582) [reply stdlib/public/SwiftShims/LibcShims.h]: Changed to `__swift_bool`. 

Any particular reason for the `!= 0`?

File: stdlib/public/core/Shims.swift
Comment (3405691582) [reply stdlib/public/core/Shims.swift]: Done.

File: stdlib/public/SwiftShims/LibcShims.h
Comment (benrimmington) [reply stdlib/public/SwiftShims/LibcShims.h]: > Any particular reason for the `!= 0`?

No reason. I think the implicit conversion is good for `__swift_bool` (aka C++ `bool` or C99 `_Bool`). But it would be bad for Objective-C [`BOOL`](https://developer.apple.com/documentation/objectivec/bool?language=objc#discussion) (aka `signed char` on macOS).

File: stdlib/public/SwiftShims/LibcShims.h
Comment (benrimmington) [inline stdlib/public/SwiftShims/LibcShims.h:118]: Does this `#define HAS_MALLOC_SIZE` twice for Android?

Comment (benrimmington) [reply stdlib/public/SwiftShims/LibcShims.h:118]: Also, should `HAS_MALLOC_SIZE` be renamed to `_SWIFT_STDLIB_HAS_MALLOC_SIZE`?

File: stdlib/public/SwiftShims/LibcShims.h
Comment (3405691582) [reply stdlib/public/SwiftShims/LibcShims.h]: OK, added it.

File: stdlib/public/SwiftShims/LibcShims.h
Comment (3405691582) [reply stdlib/public/SwiftShims/LibcShims.h]: Good catch. Removed the Android #define.

I'm not sure a rename is necessary; giving it a full prefix suggests this is something with stronger scope than it actually has. The underscore prefix seems rather uncommon for stdlib as well, so I'm inclined to leave it as-is.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by benrimmington** [comment]
The macOS failure was:

```
dyld: Symbol not found: _$ss11_mallocSize12ofAllocationSiSgSV_tF
  Referenced from: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swift-test
  Expected in: /usr/lib/swift/libswiftCore.dylib
```

Does this mean that the *system* dylib is expected to contain the new `_mallocSize(ofAllocation:)` function, when building SwiftPM?

**Comment by compnerd** [comment]
Yes, the tests currently run against the system version of the core library.  This is due to a system issue that’s been there for a while.  Fixing that is blocked by the other build system work that exposes the issue in an addressable manner.

CC: @drexin

**Comment by jckarter** [comment]
Running tests against the OS runtime seems like a feature, not a bug, because we want to know if we accidentally add symbols to the standard library, since they'll be burdened by deployment target limitations because they aren't available in older OSes.

In this case, we need to mark that new function with the `@_alwaysEmitIntoClient` attribute, so that it's not part of the standard library ABI.

File: stdlib/public/core/Shims.swift
Comment (jckarter) [inline stdlib/public/core/Shims.swift:40]: In addition to being `@usableFromInline`, this should also be `@_alwaysEmitIntoClient`, so that user code doesn't try to link against a nonexistent symbol in older Swift runtime dylibs.

File: stdlib/public/SwiftShims/LibcShims.h
Comment (benrimmington) [reply stdlib/public/SwiftShims/LibcShims.h]: Suggestion: replace each `#define HAS_MALLOC_SIZE` with a function?

```diff
-#define HAS_MALLOC_SIZE 1
+static inline __swift_bool _swift_stdlib_has_malloc_size() { return true; }
```

```diff
-#define HAS_MALLOC_SIZE 0
+static inline __swift_bool _swift_stdlib_has_malloc_size() { return false; }
```

File: stdlib/public/core/Shims.swift
Comment (3405691582) [reply stdlib/public/core/Shims.swift]: Done.

Thanks, that's the piece of knowledge I was missing here.

File: stdlib/public/SwiftShims/LibcShims.h
Comment (3405691582) [reply stdlib/public/SwiftShims/LibcShims.h]: It's a lot more to copy and paste for the next new port, though.This suggests it would be far better to push the preprocessor conditionals into one definition of each function at the top level, but that would probably be better suited to do in a separate PR.

**Comment by 3405691582** [comment]
Can someone please kick-off swift-ci on Mac OS for me?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@CodaFi This look good before I merge?

**Review by CodaFi — APPROVED** [review]

**Comment by jckarter** [comment]
Thanks @3405691582!

**Pull request merged**
